File: leetcode_75/hashmap_set/determine_if_two_strings_are_close_1657.py
from collections import Counter

class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False

        k1, v1 = set(word1), list(Counter(word1).values())
        k2, v2 = set(word2), list(Counter(word2).values())

        if k1 == k2:
            if (sorted(v1) == sorted(v2)): # ! 這邊用 sorted 還是 Counter 都可以
                return True
        return False


aa = Solution()
aa.closeStrings("abbzzca", "babzzcz")


# -----------------------------------------------------------------
# 取出第一組 key-value
# first_key, first_value = list(aa.items())
# # 如果你只想取出 value
# first_value = list(aa.values())
# # 如果你只想取出 key
# first_key = list(aa.keys())

File: leetcode_75/hashmap_set/determine_if_two_strings_are_close_1657.java
import java.util.*;

// ! 這是一個java表達法，可以直接減掉字母
// 'a' - 'a' ➜ 0
// 'b' - 'a' ➜ 1
// 'z' - 'a' ➜ 25

// 下面的更快，
// ---------------------------------------------------------------
class Solution {
    public boolean closeStrings(String word1, String word2) {
        int[] arr1 = new int[26];
        int[] arr2 = new int[26];

        for (int i = 0; i < word1.length(); i++) {
            arr1[word1.charAt(i) - 'a']++;
        }

        for (int i = 0; i < word2.length(); i++) {
            arr2[word2.charAt(i) - 'a']++;
        }

        for (int i = 0; i < 26; i++) {
            if ((arr1[i] == 0 && arr2[i] != 0) || (arr2[i] == 0 && arr1[i] != 0)) {
                // 這邊的第一重辨別，檢查一邊為 0 但另一邊不是，並且兩邊檢查
                // 不能直接雙邊等價的檢查，因為我們還需要置換數字，不是單純判斷等價
                return false;
            }
        }

        Arrays.sort(arr1);
        Arrays.sort(arr2);

        return Arrays.equals(arr1, arr2);
    }
}

// ---------------------------------------------------------------v
class Solution {
    public boolean closeStrings(String word1, String word2) {
        if (word1.length() != word2.length())
            return false;

        // 建立兩個 Counter
        Map<Character, Integer> map1 = countChars(word1);
        Map<Character, Integer> map2 = countChars(word2);

        // 1. 比較 key 的集合是否相同
        if (!map1.keySet().equals(map2.keySet()))
            return false;

        // 2. 比較 value 的頻率（不在乎哪個字母對應哪個數量）
        List<Integer> freq1 = new ArrayList<>(map1.values());
        List<Integer> freq2 = new ArrayList<>(map2.values());
        Collections.sort(freq1);
        Collections.sort(freq2);

        return freq1.equals(freq2);
    }

    // 模擬 Counter 的方法
    private Map<Character, Integer> countChars(String word) {
        Map<Character, Integer> map = new HashMap<>();
        for (char c : word.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }
        return map;
    }
}
